processing/src/main/java/org/apache/druid/query/planning/ExecutionVertex.java

```diff
@@ -0,0 +1,311 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.planning;
+
+import com.google.common.base.Preconditions;
+import org.apache.druid.error.DruidException;
+import org.apache.druid.java.util.common.Intervals;
+import org.apache.druid.query.BaseQuery;
+import org.apache.druid.query.DataSource;
+import org.apache.druid.query.Druids;
+import org.apache.druid.query.JoinDataSource;
+import org.apache.druid.query.Query;
+import org.apache.druid.query.QueryDataSource;
+import org.apache.druid.query.TableDataSource;
+import org.apache.druid.query.UnionDataSource;
+import org.apache.druid.query.scan.ScanQuery;
+import org.apache.druid.query.spec.MultipleIntervalSegmentSpec;
+import org.apache.druid.query.spec.QuerySegmentSpec;
+import org.apache.druid.segment.join.JoinPrefixUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Represents the native engine's execution vertex.
+ *
+ * Most likely due to evolutional purposes there are some concepts which went
+ * beyond their inital design:
+ *
+ * Multiple queries might be executed in one stage: <br/>
+ * The GroupBy query could be collapsed at exection time).
+ *
+ * Dag of datasources: <br/>
+ * an execution may process an entire dag of datasource in some cases
+ * (joindatasource) ; or collapse some into the execution (filter)
+ */
```

Comment:
The javadoc could use some improvement, such as:

- Define "execution vertex" more clearly. It can be difficult to find the right words. Examples are probably the best way to clearly explain it.
- Please don't discuss history of how we got here ("evolution") in the javadoc. That should be for git history and PR descriptions. Readers of the javadoc should be focused on the current state only.